storage/Filesystems/MountableImpl.h

```diff
@@ -0,0 +1,214 @@
+
+	virtual string get_mount_type() const = 0;
+
+	const Mountable* get_mountable() const { return to_mountable(get_device()); }
+
+	virtual Text do_mount_text(const string& mountpoint, Tense tense) const;
+	virtual void do_mount(const Actiongraph::Impl& actiongraph, const string& mountpoint) const;
+
+	virtual Text do_umount_text(const string& mountpoint, Tense tense) const;
+	virtual void do_umount(const Actiongraph::Impl& actiongraph, const string& mountpoint) const;
+
+	virtual Text do_add_to_etc_fstab_text(const string& mountpoint, Tense tense) const;
+	virtual void do_add_to_etc_fstab(const Actiongraph::Impl& actiongraph, const string& mountpoint) const;
+
+	virtual Text do_remove_from_etc_fstab_text(const string& mountpoint, Tense tense) const;
+	virtual void do_remove_from_etc_fstab(const Actiongraph::Impl& actiongraph, const string& mountpoint) const;
+
+    protected:
+
+	Impl()
+	    : Device::Impl(), mountpoints({}), mount_by(MountByType::DEVICE), fstab_options({})
+	    {}
+
+	Impl(const xmlNode* node);
+
+	void save(xmlNode* node) const override;
+
+    private:
+
+	// TODO this should be a list of a struct with mountpoint, mount-by
+	// and fstab-options. or add a mountpoint object to the devicegraph?
+	vector<string> mountpoints;
+	MountByType mount_by;
+	list<string> fstab_options;
+
+    };
+
+
+    namespace Action
+    {
+
+	class Mount : public Modify
+	{
+	public:
+
+	    Mount(sid_t sid, const string& mountpoint)
+		: Modify(sid), mountpoint(mountpoint) {}
+
+	    virtual Text text(const Actiongraph::Impl& actiongraph, Tense tense) const override;
+	    virtual void commit(const Actiongraph::Impl& actiongraph) const override;
+
+	    const string mountpoint;
+
+	};
+
+
+	class Umount : public Modify
+	{
+	public:
+
+	    Umount(sid_t sid, const string& mountpoint)
+		: Modify(sid), mountpoint(mountpoint) {}
+
+	    virtual Text text(const Actiongraph::Impl& actiongraph, Tense tense) const override;
+	    virtual void commit(const Actiongraph::Impl& actiongraph) const override;
+
+	    const string mountpoint;
+
+	};
+
+
+	class AddToEtcFstab : public Modify
+	{
+	public:
+
+	    AddToEtcFstab(sid_t sid, const string& mountpoint)
+		: Modify(sid), mountpoint(mountpoint) {}
+
+	    virtual Text text(const Actiongraph::Impl& actiongraph, Tense tense) const override;
+	    virtual void commit(const Actiongraph::Impl& actiongraph) const override;
+
+	    virtual void add_dependencies(Actiongraph::Impl::vertex_descriptor vertex,
+					  Actiongraph::Impl& actiongraph) const override;
+
+	    const string mountpoint;
+
+	};
+
+
+	class RemoveFromEtcFstab : public Modify
+	{
+	public:
+
+	    RemoveFromEtcFstab(sid_t sid, const string& mountpoint)
+		: Modify(sid), mountpoint(mountpoint) {}
+
+	    virtual Text text(const Actiongraph::Impl& actiongraph, Tense tense) const override;
+	    virtual void commit(const Actiongraph::Impl& actiongraph) const override;
+
+	    const string mountpoint;
+
+	};
+
+    }
+
```

Comment:
I'm not sure I understand the issue. Of course the interfaces look identical since they are all derived from Action::Modify. But when you reimplement a function in C++ you have to provide the declaration in the class definition so in this case even when deriving from MountableAction the text and commit functions must be declared in Mount, Umount, ...